### PR TITLE
Add tsx (TypeScript + JSX) support

### LINKED ISSUE #1155: (typescript) Add JSX support to Typescript
There's a somewhat popular variant of TypeScript that incorporates JSX, and replaces TypeScript's normal angle-bracket casts with the `as` keyword.  Source files typically end in `.tsx`.  It would be nice to have a language definition for this -- it's not hard to make one, but not clear how to cleanly share the bulk of the definition with `typescript`.

Related issues: #878, reviewable/reviewable#285

### PR DESCRIPTION
**Pull request by ddevault**
This uses the same rules as JavaScript's JSX support.

Fixes #1155

### CODE REVIEW

**Comment by zoran5** [comment]
@SirCmpwn Could you please not put the return statement in the constructor, but instead create a separate function?

**Comment by SirCmpwn** [comment]
Does it matter? This is just for testing the highlighitng kicks in right.

**Comment by zoran5** [comment]
@SirCmpwn I think the same examples are also used on https://highlightjs.org/static/demo/  
And I'm pretty sure the users who would see that typescript code there would get annoyed.

**Comment by mmartinson** [comment]
Anything I can do to help move this one along? Testing, changes?

**Comment by SirCmpwn** [comment]
https://github.com/isagalaev/highlight.js/issues/1678

**Comment by paladox** [comment]
Hi, this needs rebasing.

**Comment by paladox** [comment]
@ddevault would you be able to rebase please?

**Comment by rhigdon** [comment]
https://github.com/highlightjs/highlight.js/pull/1966
I rebased Drew's patch and updated the CHANGES.md and AUTHORS.en.txt

**Comment by PabloSzx** [comment]
any news on this?

File: test/markup/typescript/jsx.expect.txt [hunk 1]
```diff
@@ -33,9 +33,10 @@
         <span class="hljs-keyword">return</span> state;
     }
     <span class="hljs-keyword">else</span> <span class="hljs-keyword">if</span> (node.kind === SyntaxKind.ModuleDeclaration) {
-        <span class="hljs-keyword">return</span> getModuleInstanceState((&lt;ModuleDeclaration&gt;node).body);
+        <span class="hljs-keyword">return</span> getModuleInstanceState((<span class="xml"><span class="hljs-tag">&lt;<span class="hljs-name">ModuleDeclaration</span>&gt;</span>node).body);
     }
-    <span class="hljs-keyword">else</span> {
```
Comment (yyyc514) [inline test/markup/typescript/jsx.expect.txt]: Why is the syntax highlighting on else and return breaking here?

File: test/detect/typescript/default.txt [hunk 2]
```diff
@@ -2,6 +2,7 @@ class MyClass {
   public static myValue: string;
   constructor(init: string) {
     this.myValue = init;
+    return <div className="test">jsx syntax</div>;
```
Comment (yyyc514) [inline test/detect/typescript/default.txt]: Please also add a test with a self-closing tag as that seems to be a big issue with our React/JSX matching feature for this...

**Comment by yyyc514** [comment]
Closing this as it's out of date now. 

I think we're open to this, but it should be re-reviewed after 10.0 when we can make a single tiny JSX module and then import that exact module into both JS and TS, otherwise this is going to diverge and get very messy over time (as copy and paste tends to do).

**Pull request closed without merging**